File: Data Structures & Algorithms/search-2d-matrix/submission-0.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:

        arr = [i for m in matrix for i in m]
        size = len(arr) - 1

        def bisearch(arr, target, left, right):
            if (right >= left):
                mid = left + (right-left)//2

                if (arr[mid] == target):
                    return True
                elif (arr[mid] > target):
                    return bisearch(arr, target, left, mid-1)
                else:
                    return bisearch(arr, target, mid+1, right)

            else:
                return False


        res = bool(bisearch(arr, target, 0, size))

        return res

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-0.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
        int len = s.length();

        if (len == 1) {
            return 1;
        }

        int arr[26] = {0};
        int left = 0, max_length = 0, max_count = 0;
        
        for (int right = 0; right < len; right++) {
            arr[s[right]-'A']++;
            max_count = max(max_count, arr[s[right]-'A']);

            while ((right-left+1) - max_count > k) {
                arr[s[left]-'A']--;
                left++;
            }

            max_length = max(max_length, right-left+1);
        }

        return max_length;
    }
};
